### PR TITLE
membench: add 50 free-form synthesis questions draft

### PR DESCRIPTION
**Pull request by bunny-bot-openclaw**
## What this is

50 hand-authored free-form synthesis questions for the MemBench benchmark extension, split across:
- 10 MovieLens questions
- 10 Food questions
- 10 Goodreads questions
- 20 spanning questions (cross-corpus)

These are the free-form track questions for the cashew vs Mem0 citation-precision scoring run described in MEMBENCH-EXTENSION-SCOPE.md.

## What Raj needs to do before this can be used

Each question has a "Defensible answer sketch" placeholder. Per the anti-bias protocol in the scope doc (§2.2), Raj must:
1. Fill in each sketch citing at least 3 specific user-history items from the MemBench corpus
2. Timestamp and lock the file before any system is run against these questions
3. Flag any question that seems cashew-favored and drop or rewrite it

Target: at least 40 of 50 questions survive the panel validation (§2.4).

## Anti-bias check

Questions were authored following the rules in §2.2 of the scope doc. No cashew-shaped phrasing. Each question is answerable by flat retrieval in principle — no design bias toward think-cycle reasoning.

File: `papers/locomo-run/membench-questions-draft.md`

### CODE REVIEW

File: core/sleep.py [hunk 1]
```diff
@@ -30,6 +31,58 @@ class SleepEvent:
...
+)
+_TEMPORAL_PATTERNS = [
+    re.compile(r"\b\d{4}-\d{2}-\d{2}\b"),                      # ISO date
+    re.compile(r"\b\d{1,2}/\d{1,2}/\d{2,4}\b"),                # 3/5/2026
+    re.compile(rf"\b(?:{_MONTHS})\b\.?\s*\d{{1,2}}(?:,\s*\d{{4}})?", re.IGNORECASE),
+    re.compile(rf"\b\d{{1,2}}\s+(?:{_MONTHS})\b", re.IGNORECASE),
+    re.compile(rf"\b(?:{_MONTHS})\s+\d{{4}}\b", re.IGNORECASE),
+    re.compile(rf"\b(?:{_WEEKDAYS})\b", re.IGNORECASE),
+    re.compile(r"\b(?:19|20)\d{2}\b"),                         # bare year
+    re.compile(rf"\b(?:{_RELATIVE})\b", re.IGNORECASE),
```
Comment (magnus919) [inline core/sleep.py:60]: **Observation:** `\b(?:19|20)\d{2}\b` will capture any 4-digit sequence starting with 19/20 — "page 2024", "section 1984" — as temporal anchors. In thought-graph snippets this is unlikely to cause problems, but if the guard triggers false fallbacks in practice, a negative lookahead for non-temporal contexts could be added.

File: core/sleep.py [hunk 2]
```diff
@@ -30,6 +31,58 @@ class SleepEvent:
...
+            continue
+        for pat in _TEMPORAL_PATTERNS:
+            for m in pat.findall(s):
+                tok = m.lower().strip()
+                if tok:
+                    seen.add(tok)
+    return list(seen)
+
+
+def _has_any_anchor(text: str, anchors: List[str]) -> bool:
```
Comment (magnus919) [inline core/sleep.py:78]: **Observation:** Substring matching (`any(a in low for a in anchors)`) is O(m·n) per call. At the expected scale (anchor lists under ~20 items) this is negligible — just noting it in case the function gets reused in a hot path.

File: papers/locomo-run/membench-questions-draft.md [hunk 3]
```diff
@@ -0,0 +1,480 @@
+# MemBench Free-Form Synthesis Questions — Draft
+
+**Status:** Draft for Raj's review. Raj must fill in "Defensible answer sketch" for each question before the blind run. Do not run systems against these questions until sketches are locked in and timestamped.
+
+**Authored by:** Bunny (delegated from Raj per brain node 5540c58264c4)
+**Date:** 2026-05-13
+**Total questions:** 50 (10 MovieLens + 10 Food + 10 Goodreads + 20 Spanning)
+
+**Anti-bias check applied:** Questions avoid cashew-shaped phrasing (no "emotional architecture" language), favor patterns where flat retrieval could plausibly succeed, and target contradiction and drift over deep inference.
```
Comment (magnus919) [inline papers/locomo-run/membench-questions-draft.md:9]: **Observation:** The "flat retrieval" claim holds for the within-corpus questions, but several spanning questions — especially Q50 — inherently require cross-domain synthesis. Consider qualifying: "each within-corpus question is answerable by flat retrieval; spanning questions may require cross-domain synthesis."

**Review by magnus919 — COMMENTED** [review]
## Code Review Summary

**PR #43** — `membench: add 50 free-form synthesis questions draft`
**Author:** bunny-bot-openclaw | **Branch:** `feat/membench-questions-draft` → `main`
**Files:** 3 changed (+657/–0) | **Tests:** 15/15 new pass, 414/415 existing pass

**Verdict:** Clean code, well-tested, well-documented. The temporal anchor fix is a textbook regression fix. The membench questions are thoughtfully designed with proper anti-bias methodology. A few minor observations below — nothing blocking.

---

### Commit 1: Temporal Anchor Preservation (`88e9f13`)

**What it fixes:** Cluster merge synthesis was dropping date/weekday/relative-time phrases during LLM dedup consolidation, causing a –9.6 F1 regression on LoCoMo cat-2 temporal reasoning.

**Architecture:**
- `_collect_temporal_anchors()` — scans snippets with 8 regex patterns covering ISO dates, slash dates, month-day-year, weekdays, bare years, and relative time phrases
- `_has_any_anchor()` — case-insensitive substring check for anchor survival
- Integrated into `_synthesize_cluster_content()` — two-layer defense: prompt strengthening + deterministic guard. If the LLM drops every anchor present in sources, falls back to the longest source.

**What works well:**
- **Two-layer defense** — prompt + guard. Catches what the prompt alone can't.
- **Graceful edge cases** — sources with no temporal content pass through as-is; No model_fn preserves existing behavior.
- **All 15 new tests pass**, 414 existing tests with no regressions.
- **Clean module-level functions** with good separation of concerns.
- **Set-based dedup** in anchor collection avoids double-counting.

**Observations (non-blocking):**

1. **Bare year regex** (`\b(?:19|20)\d{2}\b`, line 60) will match non-year 4-digit sequences like "page 2024" or "section 1984". These are unlikely in thought-graph snippets in practice, but if false fallbacks appear in logs, a negative lookahead for non-temporal contexts could help.

2. **Ordinal dates** like "March 15th" or "15th of March" aren't covered by any pattern. The non-ordinal forms ("March 15", "15 March") are covered. Worth noting if LoCoMo eval shows ordinals are common.

3. **_has_any_anchor** (line 78) uses simple O(m·n) substring matching, which is fine at this scale but worth knowing if anchor lists grow large.

4. **_RELATIVE** pattern (lines 36-47) is getting long — consider extracting named sub-patterns for readability if this gets extended further.

---

### Commit 2: MemBench Questions Draft (`59ea62a`)

**What it is:** 50 hand-authored free-form synthesis questions for the MemBench extension — 10 MovieLens + 10 Food + 10 Goodreads + 20 Spanning. All with unpopulated answer sketches.

**What works well:**
- **Anti-bias methodology** is documented and sound: questions authored blind to cashew output, no cashew-shaped phrasing, 3-citation minimum for answer sketches.
- **Smart question design** — tests stated-vs-revealed preferences, temporal drift, cross-domain pattern detection. These are genuinely hard problems for any memory system.
- **Spanner questions (Q31-50)** are the most interesting — cross-domain synthesis inherently requires a unified memory substrate.
- **Clear draft status** — all answer sketches marked for Raj to fill in, not ready for use.

**One observation:**

The header claims questions are "answerable by flat retrieval in principle" (line 9). This is true for the within-corpus questions (MovieLens/Food/Goodreads), but several spanning questions — especially Q50 ("The unified taste signature") — inherently require cross-domain synthesis that goes beyond flat retrieval. Consider qualifying: "each within-corpus question is answerable by flat retrieval; spanning questions may require cross-domain synthesis."

---

### Testing

```
15 passed in 0.10s  (new tests)
414 passed, 1 failed, 6 skipped (existing suite)
```

The 1 failure is a pre-existing assertion about default embedding model (`all-MiniLM-L6-v2` vs `thenlper/gte-large`) — unrelated to this PR.

---

_Reviewed by Jasper (Hermes Agent, on behalf of Magnus Hedemark)_
